File: backends/ts/src/keyframe-expression.ts
import type { EasingKind, Keyframe, KeyframeProp } from "@aicut/core";

/**
 * Compile a per-property keyframe array into an ffmpeg filter
 * expression. The result is suitable to drop into
 * `scale=w='...':h='...':eval=frame` or `overlay=x='...':eval=frame`
 * — both filters expose `t` as the current OUTPUT timestamp.
 *
 * Math: each segment between kf[i] and kf[i+1] contributes
 *
 *   gte(T, A) * lt(T, B) * ( a.value + easedT * (b.value - a.value) )
 *
 * where `T` is the time-variable expression (default `t`) and
 * easedT applies the LEAVING keyframe's outgoing curve to the
 * normalized progress `(T - A)/(B - A)`. Before-first holds the
 * first value; after-last holds the last value. Boundaries use a
 * half-open `[A, B)` mask so the same T never gets counted twice.
 *
 * For props with NO keyframes the function returns the static
 * fallback as a literal so the caller can drop the result straight
 * into the filter without conditionals.
 *
 * `tVar` lets the caller substitute a CLIP-LOCAL expression when the
 * filter sees global timeline `t` — e.g. when compositing multiple
 * tracks in a single filter_complex pass, callers pass
 * `tVar: "(t-${clip.start/1000})"` so keyframe times stay anchored
 * to clip start.
 */
export interface CompileOpts {
  /** Expression that evaluates to clip-local seconds. Defaults to `"t"`. */
  tVar?: string;
}

export function compileKeyframeExpression(
  keyframes: Keyframe[] | undefined,
  prop: KeyframeProp,
  fallback: number,
  opts: CompileOpts = {},
): string {
  const T = opts.tVar ?? "t";
  if (!keyframes || keyframes.length === 0) return formatNumber(fallback);
  const propKfs = keyframes
    .filter((k) => k.prop === prop)
    .sort((a, b) => a.time - b.time);
  if (propKfs.length === 0) return formatNumber(fallback);
  if (propKfs.length === 1) return formatNumber(propKfs[0]!.value);

  const parts: string[] = [];
  const first = propKfs[0]!;
  const last = propKfs[propKfs.length - 1]!;

  // Before first kf — held at first.value.
  parts.push(`lt(${T},${formatTime(first.time)})*${formatNumber(first.value)}`);

  // Segments.
  for (let i = 0; i < propKfs.length - 1; i += 1) {
    const a = propKfs[i]!;
    const b = propKfs[i + 1]!;
    const aSec = formatTime(a.time);
    const bSec = formatTime(b.time);
    const dur = (b.time - a.time) / 1000;
    if (dur <= 0) continue; // zero-length segment — skip to avoid div/0
    const rawT = `(${T}-${aSec})/${formatNumber(dur)}`;
    const easedT = easingExpression(rawT, a.easing ?? "linear");
    const delta = b.value - a.value;
    const lerpedExpr =
      delta === 0
        ? formatNumber(a.value) // value held — constant segment
        : `(${formatNumber(a.value)}+(${easedT})*(${formatNumber(delta)}))`;
    parts.push(`gte(${T},${aSec})*lt(${T},${bSec})*${lerpedExpr}`);
  }

  // After last kf — held at last.value.
  parts.push(`gte(${T},${formatTime(last.time)})*${formatNumber(last.value)}`);

  return parts.join("+");
}

/**
 * Wrap a `t∈[0,1]` ffmpeg expression in the easing curve from our
 * frontend. Matches `applyEasing` in `keyframes/interpolate.ts` line
 * for line — same cubic formulas. Endpoints (rawT=0 and rawT=1)
 * always evaluate to 0 and 1 so the kf values stay exact.
 *
 * `rawT` MUST be a parenthesized expression — callers feed
 * `(t-A)/D` style, never a single variable. The output is also
 * safe to nest inside other expressions because the body wraps
 * with explicit parens.
 */
export function easingExpression(rawT: string, easing: EasingKind): string {
  switch (easing) {
    case "linear":
      return rawT;
    case "easeIn":
      return `pow(${rawT},3)`;
    case "easeOut":
      return `(1-pow(1-(${rawT}),3))`;
    case "easeInOut":
      // Same cubic split as the frontend: t<0.5 → 4t³, else 1-(-2t+2)³/2.
      return `if(lt(${rawT},0.5),4*pow(${rawT},3),1-pow(-2*(${rawT})+2,3)/2)`;
  }
}

/** ffmpeg-friendly number — 6 sig figs of decimals, no exponential. */
function formatNumber(n: number): string {
  if (Number.isInteger(n)) return String(n);
  // Avoid scientific notation entirely; ffmpeg's parser doesn't grok it.
  const s = n.toFixed(6);
  // Trim trailing zeros (and a stranded `.`) so the expression
  // stays compact and grep-friendly.
  return s.replace(/0+$/, "").replace(/\.$/, "");
}

/** Convert clip-local ms to seconds at ffmpeg-expression precision. */
function formatTime(ms: number): string {
  return formatNumber(ms / 1000);
}

File: backends/go/keyframe_expression.go
package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// compileKeyframeExpression returns an ffmpeg filter expression in `t`
// (output-stream seconds) that evaluates to the property's value at
// time t. Mirrors backends/ts/src/keyframe-expression.ts line for
// line — same math, same boundary semantics:
//
//   - No keyframes for the prop → return the fallback as a literal.
//   - Single keyframe        → its value, constant.
//   - N keyframes            → sum of per-segment terms, each masked
//     by `gte(t, A) * lt(t, B)` so the half-open window [A, B) covers
//     exactly one segment per t.
//   - Before-first / after-last → held at the boundary kf's value.
//
// Easing follows the LEAVING keyframe (a→b is shaped by a.easing),
// matching the frontend convention.
func compileKeyframeExpression(keyframes []Keyframe, prop string, fallback float64) string {
	if len(keyframes) == 0 {
		return formatNumber(fallback)
	}
	propKfs := make([]Keyframe, 0, len(keyframes))
	for _, k := range keyframes {
		if k.Prop == prop {
			propKfs = append(propKfs, k)
		}
	}
	if len(propKfs) == 0 {
		return formatNumber(fallback)
	}
	sort.SliceStable(propKfs, func(i, j int) bool { return propKfs[i].Time < propKfs[j].Time })
	if len(propKfs) == 1 {
		return formatNumber(propKfs[0].Value)
	}

	first := propKfs[0]
	last := propKfs[len(propKfs)-1]

	parts := make([]string, 0, len(propKfs)+1)

	// Before first kf — held at first.value.
	parts = append(parts, fmt.Sprintf("lt(t,%s)*%s", formatTime(first.Time), formatNumber(first.Value)))

	// Segments.
	for i := 0; i < len(propKfs)-1; i++ {
		a := propKfs[i]
		b := propKfs[i+1]
		if b.Time <= a.Time {
			continue // zero / negative duration — skip to avoid div-by-zero
		}
		aSec := formatTime(a.Time)
		bSec := formatTime(b.Time)
		dur := float64(b.Time-a.Time) / 1000.0
		rawT := fmt.Sprintf("(t-%s)/%s", aSec, formatNumber(dur))
		easedT := easingExpression(rawT, a.Easing)
		delta := b.Value - a.Value
		var lerpedExpr string
		if delta == 0 {
			lerpedExpr = formatNumber(a.Value)
		} else {
			lerpedExpr = fmt.Sprintf("(%s+(%s)*(%s))", formatNumber(a.Value), easedT, formatNumber(delta))
		}
		parts = append(parts, fmt.Sprintf("gte(t,%s)*lt(t,%s)*%s", aSec, bSec, lerpedExpr))
	}

	// After last kf — held at last.value.
	parts = append(parts, fmt.Sprintf("gte(t,%s)*%s", formatTime(last.Time), formatNumber(last.Value)))

	return strings.Join(parts, "+")
}

// easingExpression wraps a normalized `rawT∈[0,1]` ffmpeg expression
// in the easing curve from the frontend. Endpoints stay exact (0→0,
// 1→1) so kf positions don't drift. Unknown easing values fall back
// to linear — defensive parse of project JSON authored elsewhere.
func easingExpression(rawT string, easing string) string {
	switch easing {
	case "easeIn":
		return fmt.Sprintf("pow(%s,3)", rawT)
	case "easeOut":
		return fmt.Sprintf("(1-pow(1-(%s),3))", rawT)
	case "easeInOut":
		return fmt.Sprintf("if(lt(%s,0.5),4*pow(%s,3),1-pow(-2*(%s)+2,3)/2)", rawT, rawT, rawT)
	default:
		// "linear" or anything we don't recognise.
		return rawT
	}
}

// formatNumber returns the ffmpeg-friendly decimal form of n. Integers
// stay integer-printed, fractions are truncated to 6 sig digits and
// stripped of trailing zeros / the stranded dot. Avoids scientific
// notation, which ffmpeg's parser doesn't understand.
func formatNumber(n float64) string {
	if n == float64(int64(n)) {
		return strconv.FormatInt(int64(n), 10)
	}
	s := strconv.FormatFloat(n, 'f', 6, 64)
	// Trim trailing zeros, then a possibly-stranded dot.
	s = strings.TrimRight(s, "0")
	s = strings.TrimRight(s, ".")
	return s
}

// formatTime converts clip-local ms to ffmpeg expression seconds.
func formatTime(ms int64) string {
	return formatNumber(float64(ms) / 1000.0)
}

// floatOr returns *p if non-nil, else fallback. Helper for the
// pointer-typed PanX / PanY / Scale fields on Clip.
func floatOr(p *float64, fallback float64) float64 {
	if p == nil {
		return fallback
	}
	return *p
}
